File: frontends/nextjs/src/web/features/auth/domain/validation.test.ts
import { describe, expect, it } from "vitest";

import { validateLoginForm, validateRegisterForm } from "./validation";

describe("validateRegisterForm", () => {
  it("accepts a genuinely valid submission", () => {
    const errors = validateRegisterForm({
      email: "fatima.rahman@example.com",
      password: "correct-horse-battery",
      displayName: "Fatima Rahman",
    });
    expect(errors).toEqual({});
  });

  it("rejects a password under 8 characters, matching the contract's own RegisterRequest.password minLength", () => {
    const errors = validateRegisterForm({ email: "a@example.com", password: "short", displayName: "A" });
    expect(errors.password).toBeDefined();
  });

  it("rejects a malformed email", () => {
    const errors = validateRegisterForm({ email: "not-an-email", password: "correct-horse-battery", displayName: "A" });
    expect(errors.email).toBeDefined();
  });

  it("rejects an empty display name", () => {
    const errors = validateRegisterForm({ email: "a@example.com", password: "correct-horse-battery", displayName: "   " });
    expect(errors.displayName).toBeDefined();
  });

  it("rejects a display name over 200 characters, matching the contract's own maxLength", () => {
    const errors = validateRegisterForm({ email: "a@example.com", password: "correct-horse-battery", displayName: "x".repeat(201) });
    expect(errors.displayName).toBeDefined();
  });
});

describe("validateLoginForm", () => {
  it("accepts a valid submission", () => {
    expect(validateLoginForm({ email: "a@example.com", password: "anything" })).toEqual({});
  });

  it("rejects an empty password without guessing a minimum length — the backend, not this form, owns that rule for login", () => {
    const errors = validateLoginForm({ email: "a@example.com", password: "" });
    expect(errors.password).toBeDefined();
  });
});

File: mobile/flutter/test/features/auth/domain/validation_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:stackbraid_mobile/features/auth/domain/validation.dart';

void main() {
  group('validateRegisterForm', () {
    test('accepts a genuinely valid submission', () {
      final errors = validateRegisterForm(
        const RegisterFormValues(
          email: 'fatima.rahman@example.com',
          password: 'correct-horse-battery',
          displayName: 'Fatima Rahman',
        ),
      );
      expect(errors, isEmpty);
    });

    test('rejects a password under 8 characters, matching the contract\'s own RegisterRequest.password minLength', () {
      final errors = validateRegisterForm(
        const RegisterFormValues(email: 'a@example.com', password: 'short', displayName: 'A'),
      );
      expect(errors['password'], 'passwordTooShort');
    });

    test('rejects a malformed email', () {
      final errors = validateRegisterForm(
        const RegisterFormValues(email: 'not-an-email', password: 'correct-horse-battery', displayName: 'A'),
      );
      expect(errors['email'], 'emailInvalid');
    });

    test('rejects an empty display name', () {
      final errors = validateRegisterForm(
        const RegisterFormValues(email: 'a@example.com', password: 'correct-horse-battery', displayName: '   '),
      );
      expect(errors['displayName'], 'displayNameRequired');
    });

    test('rejects a display name over 200 characters, matching the contract\'s own maxLength', () {
      final errors = validateRegisterForm(
        RegisterFormValues(email: 'a@example.com', password: 'correct-horse-battery', displayName: 'x' * 201),
      );
      expect(errors['displayName'], 'displayNameTooLong');
    });
  });

  group('validateLoginForm', () {
    test('accepts a valid submission', () {
      expect(validateLoginForm(const LoginFormValues(email: 'a@example.com', password: 'anything')), isEmpty);
    });

    test('rejects an empty password without guessing a minimum length — the backend, not this form, owns that rule for login', () {
      final errors = validateLoginForm(const LoginFormValues(email: 'a@example.com', password: ''));
      expect(errors['password'], 'passwordRequired');
    });
  });
}
